Robin's Changes between RC2 and RC3

clanmills opened the pull request:
I'm rather surprised that conversions.sh reveals a (harmless) order change on:

```
runTest exiv2 -M'set Xmp.xmp.ModifyDate 2015-04-17T18:10:22Z' v.xmp
```
No other part of the test suite notices any change and the order of the Xmp keys is not important.

I have stepped the code in the debugger and the key is added in xmp.cpp:
```
    int XmpData::add(const Xmpdatum& xmpDatum)
    {
        xmpMetadata_.push_back(xmpDatum);
        return 0;
    }
```

clanmills committed:
Fix: https://github.com/Exiv2/exiv2/issues/521
conversions.sh reveals a (harmless) order change runTest exiv2 -M'set Xmp.xmp.ModifyDate 2015-04-17T18:10:22Z' v.xmp

clanmills commented:
I'm going to keep working on this until I understand exactly what happened to conversions.sh.  I've reproduced this behaviour with a simpler script **acon.sh**

```
#!/usr/bin/env bash

##
# generate the file u.xmp from "empty"
cp    ../test/data/exiv2-empty.jpg                                       u.jpg
bin/exiv2 -M'set Exif.Photo.DateTimeOriginal 2003:12:14 12:01:44'        u.jpg
bin/exiv2 -M'set Exif.Photo.SubSecTimeOriginal 999999999'                u.jpg
bin/exiv2 -M'set Exif.Photo.ExifVersion 48 50 50 49'                     u.jpg
bin/exiv2 -M'set Exif.Photo.ComponentsConfiguration 1 2 3 0'             u.jpg
bin/exiv2 -M'set Exif.Photo.Flash 73'                                    u.jpg
bin/exiv2 -M'set Exif.GPSInfo.GPSLatitude 3/1 8/1 29734512/1000000'      u.jpg
bin/exiv2 -M'set Exif.GPSInfo.GPSLatitudeRef N'                          u.jpg
bin/exiv2 -M'set Exif.GPSInfo.GPSVersionID 2 2 0 1'                      u.jpg
bin/exiv2 -M'set Exif.GPSInfo.GPSTimeStamp 1/1 2/1 999999999/1000000000' u.jpg
bin/exiv2 -eX --verbose --force u.jpg

##
# generate v.jpg from empty and then inserting u.xmp!
cp u.xmp v.xmp
cp ../test/data/exiv2-empty.jpg v.jpg
bin/exiv2 -M'set Xmp.xmp.ModifyDate 2015-04-17T18:10:22Z' v.xmp
TZ=GMT-8 bin/exiv2  -iX v.jpg
bin/exiv2 --verbose -px v.jpg

# That's all Folks!
##
```

```
677 rmills@rmillsmm:~/gnu/github/exiv2/master/build $ ../../exiv2/cmake-build-debug/acon.sh 
File 1/1: u.jpg
Writing Exif data from u.jpg to ./u.xmp
File 1/1: v.jpg
Xmp.exif.ExifVersion                         XmpText     4  2.21
Xmp.exif.GPSVersionID                        XmpText     7  2.2.0.1
Xmp.exif.GPSLatitude                         XmpText    12  3,8.4955752N
Xmp.exif.GPSTimeStamp                        XmpText    29  2003-12-14T01:02:00.999999999
Xmp.exif.Flash                               XmpText     0  type="Struct"
Xmp.exif.Flash/exif:Fired                    XmpText     4  True
Xmp.exif.Flash/exif:Return                   XmpText     1  0
Xmp.exif.Flash/exif:Mode                     XmpText     1  1
Xmp.exif.Flash/exif:Function                 XmpText     5  False
Xmp.exif.Flash/exif:RedEyeMode               XmpText     4  True
Xmp.exif.ComponentsConfiguration             XmpSeq      4  YCbCr
Xmp.xmp.ModifyDate                           XmpText    20  2015-04-17T18:10:22Z
Xmp.photoshop.DateCreated                    XmpText    29  2003-12-14T12:01:44.999999999
678 rmills@rmillsmm:~/gnu/github/exiv2/master/build $ cd ../../exiv2/cmake-build-debug/
679 rmills@rmillsmm:~/gnu/github/exiv2/exiv2/cmake-build-debug $ ../../exiv2/cmake-build-debug/acon.sh 
File 1/1: u.jpg
Writing Exif data from u.jpg to ./u.xmp
File 1/1: v.jpg
Xmp.xmp.ModifyDate                           XmpText    20  2015-04-17T18:10:22Z
Xmp.exif.ExifVersion                         XmpText     4  2.21
Xmp.exif.GPSVersionID                        XmpText     7  2.2.0.1
Xmp.exif.GPSLatitude                         XmpText    12  3,8.4955752N
Xmp.exif.GPSTimeStamp                        XmpText    29  2003-12-14T01:02:00.999999999
Xmp.exif.ComponentsConfiguration             XmpSeq      4  YCbCr
Xmp.exif.Flash                               XmpText     0  type="Struct"
Xmp.exif.Flash/exif:Fired                    XmpText     4  True
Xmp.exif.Flash/exif:Return                   XmpText     1  0
Xmp.exif.Flash/exif:Mode                     XmpText     1  1
Xmp.exif.Flash/exif:Function                 XmpText     5  False
Xmp.exif.Flash/exif:RedEyeMode               XmpText     4  True
Xmp.photoshop.DateCreated                    XmpText    29  2003-12-14T12:01:44.999999999
680 rmills@rmillsmm:~/gnu/github/exiv2/exiv2/cmake-build-debug $ 
```

clanmills committed:
Revision to PR#557.  Restored XmpData::erase() and added XmpData::eraseFamily().  Restored and augumented conversions.sh

clanmills commented:
I have dug over this carefully.  I believe the earlier version of this is safe and arrives at the correct result.  However, the test suite detected a difference.

I restored XmpDatum.erase() to its previous behaviour of deleting the item at 'pos'.  I added a new API XmpDatum.eraseFamily() which deletes the item and its children.  Also add Gilles test file DSC_3079.jpg to the test suite and added a test in conversions.sh.

I think the earlier version of the code in which I modified XmpDatum.erase() to perform its previous role AND eraseFamily() is superior and logical.  However, as we are close to the end of Exiv2 v0.27, I feel there is less risk by adding XmpDatum.eraseFamily().  actions.cpp has been modified to call eraseFamily().

clanmills committed:
Bump Version to 0.27.0.3

piponazo reviewed (approved):
The changes look good to me.

I would have preferred to see the new tests in the python suite instead of adding them in the bash tests suite (`test/conversions.sh`). However, I understand that you are not as familiar with the python suite as with the bash one. We can convert the remaining bash scripts after releasing v0.27.

Is there any chance of reducing further the size of the binary file (image) by extracting the metadata into a .exv file ?

clanmills committed:
Documentation updates + release notes

- Consuming with CMake
- Debugging Exiv2
- Added "2.13 Building Exiv2 with ccache".
- Revised "Building Exiv2 with ccache" in response to Dan's feedback.
- Fixed typo in README.md
- ReleaseNotes update in preparation for RC3.

clanmills committed:
Adding output for have_regex to version.cpp

clanmills committed:
Fix for https://github.com/Exiv2/exiv2/issues/560 (eraseFamily() crashes in MSVC/Debug).

clanmills committed:
Add cmake/msvc_conan_profiles

clanmills committed:
src/version.cpp report enable_nls

clanmills committed:
Add packaging changes for CYGWIN64/CYGWIN32 and Linux32

clanmills committed:
add logs to the build bundle

clanmills committed:
Update platform/ReadMe.txt about logs.

clanmills committed:
#485 Tweak msvc output to avoid the test suite reporting an issue with conversions.sh

clanmills committed:
Fix: https://github.com/Exiv2/exiv2/issues/571

clanmills committed:
Fix: https://github.com/Exiv2/exiv2/issues/570

clanmills committed:
Change package names Platform{64|32}[Static][Debug]

clanmills committed:
packaging cleanup

clanmills committed:
Try to fix https://github.com/Exiv2/exiv2/issues/572

clanmills committed:
Correction in futils.cpp

clanmills committed:
Correction to Microsoft sample code! https://docs.microsoft.com/en-gb/windows/desktop/Services/installing-a-service

clanmills committed:
#571 Restore the original windows getProcessCode() because Microsoft's GetModuleFileName isn't effective!

piponazo commented on src/CMakeLists.txt:
```diff
@@ -191,10 +191,12 @@ if (NOT MSVC)
 
     target_link_libraries( exiv2lib PRIVATE Threads::Threads)
+else()
+    target_link_libraries( exiv2lib PRIVATE psapi ws2_32 )
```
Now we have `target_link_libraries( exiv2lib PRIVATE psapi ws2_32 )` in two different places (when **MSVC** or when `CYGWIN OR MINGW OR MSYS`). I think that it would be simpler to just have it once when `WIN32`.

piponazo replied:
We could reuse for that the already existing conditional in line 184.

clanmills commented:
Yes.  You are right.  This can be simplified.

When I was looking at the code I thought "_oh that's odd, it doesn't seem to do this for MSVC.  So how could it ever have worked?  I'll try explicitly adding it and see what happens!"_.  We now know how it works.  When building exiv2.dll, the linker does **not** worry about unsatisfied externals.  When building (static) exiv2.lib, psapi must be linked at build-time.

You're right.  Much better to do it for WIN32 as this has nothing really to do with MSVC.  And it has nothing really to do with the build-type either (static/shared).

clanmills committed:
Updating packaging for clang: Platform[{64|32}][Clang][Static][Debug]

clanmills committed:
Correction to previous commit.

piponazo commented on cmake/packaging.cmake:
```diff
@@ -1,36 +1,58 @@
-    set(CPACK_GENERATOR ZIP)  # use .zip - less likely to damage bin/exiv2.dll permissions
-else()
-    set(CPACK_GENERATOR TGZ)  # MinGW/Cygwin/Linux/MacOS-X etc use .tar.gz
+set (CC "") # Compiler
```
Few comments here. 
1) The default value of CC is empty. Is that the intended behaviour? In case the compiler is not clang, the value will remain empty.
2) The way to check for clang can be simplified (without creating variables for it):

```
    if (${CMAKE_CXX_COMPILER_ID} MATCHES "Clang")
        set (CC Clang)
    endif()
```

clanmills replied:
Yes.  That's what I want.  We'll end up with bundle names such as:

_CYGWIN64ClangStaticDebug_

However, if for a SharedRelease built with GCC, the name will simply be:

_CYGWIN64_

I think there are 60 common build possibilities.
{Linux|CYGWIN|MinGW}{64|32}{''|Clang}{''|Static}{''|Debug} = 3x2x2x2x2 = 48
{Darwin}{''|Static}{''|Debug} = 2x2 = 4
{msvc}{64|32}{''|Static}{''|Debug} = 8
Total = 60

If I decide to to build with the other versions of Visual Studio {2008|2010|2012|2013|2015), that will be 5x8=40 more builds. 

If we have a **FULL** build (webready, video) that would double it to 200.

There's no problem presenting and storing this on the buildserver.  http://exiv2.dyndns.org:8080/userContent/builds/Platform/

On the downloads page, we'll only offer 6 http://exiv2.dyndns.org/download.html
_Source_
_msvc64_
_Linux64_
_Darwin_
_MinGW64_
_CYGWIN64_

For RC1 and RC2, I published MinGW64 and MinGW32 on _'downloads'_ because we were experiencing different issues on those platforms.  However, we've put that behind us now, so I'll only offer MinGW64 with RC3.

D4N commented on README.md:
````diff
@@ -492,6 +493,33 @@ Team Exiv2 have not explored this configuration and cannot provide support.
+$ make
+```
+
+Due to the way in which ccache is installed in Fedora, this option is defeated on that platform.  [https://github.com/Exiv2/exiv2/issues/361](https://github.com/Exiv2/exiv2/issues/361)
````
This sounds like ccache does not work on Fedora, which is no true. This option just can't turn it off, that's all.

clanmills replied:
Ah @D4N Thanks for the clarification.  As I said in #361 "This is very confusing.".  I'll update the documentation tomorrow.

clanmills replied:
@piponazo I've changed the code in cmake/packaging.cmake that detects clang.

piponazo commented:
@clanmills since this PR is going to merge 0.27-RC3 into master, which is fixing many issues, could you change the description of the PR and create a list of the issues which you have fixed in the branch ? Right now, if you look at the PR description it looks like it is just fixing one issue.

clanmills commented:
Sure.  When I get to the end of this marathon, I'll update releasenotes/releasenotes.txt with a list of everything changed since RC2.  It will include:

1) Updates to documentation concerning Debugging, Clang and ccache.
2) Build and test logs are stored in the build bundles.
3) Build Server publishes and tests Cygwin, MinGW, Linux, Darwin and msvc in Release/Debug,  64/32, gcc/clang, static/shared.
4) Revised man page exiv2.1.
5) Fix del XMP/Region (#521)
6) All compiler warnings have been eliminated.
7) Web-site has been refined and polished.
8) Preparation to rehost exiv2.org
9) More security fixes (#425, #547, #558, #564)
10) Something I've forgotten!
11) Something we'll discover today or tomorrow! 

We will reach RC3 tomorrow (Friday 30 November).  Any security fixes next week will be added and RC3 will be published on Friday 7 December.  I'm hoping for no changes between RC3 and  GM.

If we have "dot" releases, every quarter,  security fixes will go into 0.27.1 at the end of March 2019.

clanmills committed:
Change how packaging detects Clang.

clanmills committed:
Restore cpack_source_generator

clanmills committed:
Do not put "Clang" in the Darwin build bundle name.
